sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,307 @@ def transform(self, X):
+        input matrix X will be overwritten with zeros.
+
+    Attributes
+    ----------
+    statistics_ : 1-D array of length {n_features}
+        The 1-D array contains the mean of each feature calculated using
+        observed (i.e. non-missing) values.
+
+    References
+    ----------
+    * Olga Troyanskaya, Michael Cantor, Gavin Sherlock, Pat Brown, Trevor
+      Hastie, Robert Tibshirani, David Botstein and Russ B. Altman, Missing
+      value estimation methods for DNA microarrays, BIOINFORMATICS Vol. 17
+      no. 6, 2001 Pages 520-525.
+
+    Examples
+    --------
+    >>> from sklearn.preprocessing.imputation import KNNImputer
+    >>> nan = float("NaN")
+    >>> X = [[1, 2, nan], [3, 4, 3], [nan, 6, 5], [8, 8, 7]]
+    >>> imputer = KNNImputer(n_neighbors=2, weights="uniform")
+    >>> imputer.fit_transform(X)
+    array([[ 1. ,  2. ,  4. ],
+           [ 3. ,  4. ,  3. ],
+           [ 5.5,  6. ,  5. ],
+           [ 8. ,  8. ,  7. ]])
+    """
+
+    def __init__(self, missing_values="NaN", n_neighbors=5,
+                 weights="uniform", metric="masked_euclidean",
+                 row_max_missing=0.5, col_max_missing=0.8, copy=True):
+
+        self.missing_values = missing_values
+        self.n_neighbors = n_neighbors
+        self.weights = weights
+        self.metric = metric
+        self.row_max_missing = row_max_missing
+        self.col_max_missing = col_max_missing
+        self.copy = copy
+
+    def fit(self, X, y=None):
+        """Fit the imputer on X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        # Imports here to avoid circular import
+        from ..neighbors import NearestNeighbors
+        from ..neighbors.base import _check_weights
+
+        # Check parameters
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=np.float64,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        self.weights = _check_weights(self.weights)
+
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed.")
+
+        # Check if % missing in any column > col_max_missing
+        mask = _get_mask(X, self.missing_values)
+        if np.any(mask.sum(axis=0) > (X.shape[0] * self.col_max_missing)):
+            raise ValueError("Some column(s) have more than {}% missing values"
+                             .format(self.col_max_missing*100))
+        X_col_means = np.ma.array(X, mask=mask).mean(axis=0).data
+
+        # Check if % missing in any row > col_max_missing
```

Comment:
Please fix this comment

Reply:
You see like a hawk my friend

Reply:
Unfortunately your reasoning is inverted: if I report a problem, it almost certainly exists; not the other way around!

Reply:
![image](https://user-images.githubusercontent.com/22569641/30577599-f6371e2c-9cd4-11e7-9c45-e6331dfce648.png)